### PR TITLE
KAFKA-6049: extend Kafka Streams Scala API for cogroup (KIP-150)

### PR DESCRIPTION
**Pull request by mjsax**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @bbejeck @vvcephei @wcarlson5

File: streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java [hunk 1]
```diff
@@ -265,15 +264,6 @@
...
 
-    /**
-     * Create a new {@link SessionWindowedCogroupedKStream} instance that can be used to perform session
-     * windowed aggregations.
-     *
-     * @param sessionWindows the specification of the aggregation {@link SessionWindows}
-     * @return an instance of {@link SessionWindowedCogroupedKStream}
-     */
-    SessionWindowedCogroupedKStream<K, VOut> windowedBy(final SessionWindows sessionWindows);
-
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/CogroupedKStream.java:276]: Just re-ordered this method to have the same ordering over multiple classes (time-window always before session-window)

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/ImplicitConversions.scala [hunk 2]
```diff
@@ -46,11 +49,24 @@ object ImplicitConversions {
...
     new KGroupedStream[K, V](inner)
 
+  implicit def wrapTimeWindowedKStream[K, V](inner: TimeWindowedKStreamJ[K, V]): TimeWindowedKStream[K, V] =
+    new TimeWindowedKStream[K, V](inner)
+
   implicit def wrapSessionWindowedKStream[K, V](inner: SessionWindowedKStreamJ[K, V]): SessionWindowedKStream[K, V] =
     new SessionWindowedKStream[K, V](inner)
 
-  implicit def wrapTimeWindowedKStream[K, V](inner: TimeWindowedKStreamJ[K, V]): TimeWindowedKStream[K, V] =
-    new TimeWindowedKStream[K, V](inner)
```
Comment (mjsax) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/ImplicitConversions.scala]: Move this one up to get better overview/grouping. The three below are new

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala [hunk 3]
```diff
@@ -525,7 +525,7 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
    * @see `org.apache.kafka.streams.kstream.KStream#join`
    * @deprecated since 2.4. Use [[KStream#join(KStream, ValueJoiner, JoinWindows, StreamJoined)]] instead.
    */
-  @deprecated
+  @deprecated("use join(KStream, ValueJoiner, JoinWindows, StreamJoined) instead", "2.4")
```
Comment (mjsax) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala]: Just a side fix -- got a warning about it

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedKStream.scala [hunk 4]
```diff
@@ -21,15 +21,18 @@ package org.apache.kafka.streams.scala
 package kstream
 
 import org.apache.kafka.streams.kstream.internals.KTableImpl
-import org.apache.kafka.streams.kstream.{KTable => KTableJ, SessionWindowedKStream => SessionWindowedKStreamJ, _}
```
Comment (mjsax) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedKStream.scala]: Only some side cleanup in this class

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedCogroupedKStream.scala [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ */
+package org.apache.kafka.streams.scala
+package kstream
+
+import org.apache.kafka.streams.kstream.{
+  SessionWindowedCogroupedKStream => SessionWindowedCogroupedKStreamJ,
+  _
+}
+import org.apache.kafka.streams.scala.FunctionsCompatConversions._
+import org.apache.kafka.streams.scala.ImplicitConversions._
```
Comment (mjsax) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedCogroupedKStream.scala]: IntelliJ complains that this is an unused import, but without it, it does not compile at command line -- what might I do wrong?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedCogroupedKStream.scala [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ */
+package org.apache.kafka.streams.scala
+package kstream
+
+import org.apache.kafka.streams.kstream.{
+  TimeWindowedCogroupedKStream => TimeWindowedCogroupedKStreamJ,
+  _
+}
+import org.apache.kafka.streams.scala.FunctionsCompatConversions._
+import org.apache.kafka.streams.scala.ImplicitConversions._
```
Comment (mjsax) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedCogroupedKStream.scala]: Same question as above.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedCogroupedKStream.scala (hunk 5 above)
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedCogroupedKStream.scala]: You can explicitly import the implicit conversions that you're using. Actually, I would recommend not even using them implicitly, but just to explicitly invoke the relevant conversion methods when you need them. (i.e., implicit conversions are syntactic sugar for users of the public API, we should try to instead write explicit and maintainable code in our internal implementations)

It should be pretty obvious which ones you need if you just delete this line and see what fails to compile.

**Review by vvcephei — COMMENTED** [review]
Hey @mjsax , thanks for adding this!

I answered your question. Otherwise, this looks reasonable, but I no longer test any Scala APIs that don't have "black box" tests... I.e., we need to write a test that uses the API just like users will to make sure it actually works the way we think it will.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedCogroupedKStream.scala (hunk 5 above)
Comment (mjsax) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedCogroupedKStream.scala]: Based on this commend, I did a proper cleanup PR: https://github.com/apache/kafka/pull/7852

We should merge the cleanup PR first, and I can rebase this PR afterwards.

**Comment by bbejeck** [comment]
@mjsax failures are relevant `Execution failed for task ':spotlessScala'.` scala formatting errors.

**Review by bbejeck — COMMENTED** [review]
Thanks for the PR @mjsax.  Overall this LGTM.  I agree with @vvcephei about adding a new test(s) for the additions.

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/ConsumedTest.scala [hunk 7]
```diff
@@ -24,8 +24,8 @@ import org.apache.kafka.streams.processor.FailOnInvalidTimestamp
 import org.apache.kafka.streams.scala.Serdes._
 import org.apache.kafka.streams.scala.Serdes
 import org.junit.runner.RunWith
-import org.scalatest.junit.JUnitRunner
 import org.scalatest.{FlatSpec, Matchers}
+import org.scalatestplus.junit.JUnitRunner
```
Comment (mjsax) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/ConsumedTest.scala:28]: Some side cleanup. `org.scalatest.junit.JUnitRunner` is deprecated

**Comment by mjsax** [comment]
Updated this PR (rebased, added tests). Call for review @bbejeck @vvcephei

**Comment by mjsax** [comment]
@vvcephei @bbejeck Rebased to resolve merge conflict. Call for review.

**Review by bbejeck — APPROVED** [review]
Thanks for the updated PR @mjsax, LGTM

**Comment by bbejeck** [comment]
Java 8 failed with 
```
kafka.admin.LeaderElectionCommandTest.testPathToJsonFile
kafka.api.PlaintextConsumerTest.testLowMaxFetchSizeForRequestAndPartition
```
Updated existing flaky test Jira tickets

Java 11 passed

retest this please

**Comment by mjsax** [comment]
Java 8 failed (one overlapping test failure):
```
kafka.admin.TopicCommandWithAdminClientTest.testCreateAlterTopicWithRackAware
kafka.api.PlaintextConsumerTest.testLowMaxFetchSizeForRequestAndPartition
```
Java 11 passed.

Retest this please.

**Comment by mjsax** [comment]
Java 11 passed.

Java 8 failed (different test):
```
org.apache.kafka.streams.integration.StandbyTaskCreationIntegrationTest.shouldCreateStandByTasksForMaterializedAndOptimizedSourceTables
```

Merging this.

**Pull request merged**